README.md
#Nagios-ElasticSearch

A few simple scripts for checking the state of your [ElasticSearch] cluster using Nagios the  endpoints of ElasticSearch

[![Build
Status](https://secure.travis-ci.org/opentable/nagios-elasticsearch.png)](https://secure.travis-ci.org/opentable/nagios-elasticsearch.png)

##How they work

These plugins work by submitting API requests to a local or remote
ElasticSearch server via the [cluster-health] and [node stats] endpoints

###Check Nodes in Cluster

This can be used to specify the nodes that are expected in the cluster. Currently, the check will return as **CRITICAL** if there are any unaccounted for nodes

```
Usage: check_es_nodes.py [options]

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -v, --verbose
  -E NODES_IN_CLUSTER, --expected_nodes_in_cluster=NODES_IN_CLUSTER
                        This is the expected number of nodes in the cluster
  -H HOST, --host=HOST  The cluster to check
  -P PORT, --port=PORT  The ES port - defaults to 9200

python check_es_nodes.py --host=myescluster.com --expected_nodes_in_cluster=13
```

###Check Cluster Health Status

This plugin can be used to get the current cluster health. This will return OK for Green, Warning for Yellow and Critical for Red.


```
Usage: check_es_cluster_status.py [options]

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -v, --verbose
  -H HOST, --host=HOST  The cluster to check
  -P PORT, --port=PORT  The ES port - defaults to 9200

python check_es_cluster_status.py --host=myescluster.com
```

###Check for Unassigned Nodes in Cluster

This plugin can be used to check for unassigned shards in the cluster. This can potentially indicate that a node is missing from the cluster. The alert will currently indicate a CRITICAL error if an unassigned shard is found

```
Usage: check_es_unassigned_shards.py [options]

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -v, --verbose
  -H HOST, --host=HOST  The cluster to check
  -P PORT, --port=PORT  The ES port - defaults to 9200


python check_es_unassigned_shards.py --host=myescluster.com
```

###Check Node JVM Usage

This plugin can be used to ensure that the nodes in the cluster are not approaching, or exceeding, the thresholds that we determine that the JVM needs to run at. Currently, if any of the nodes exceed the critical limt, then the cluster will throw a CRITICAL error. If no nodes exceed the critical threshold but a node exceeds the warning threshold, then the cluster will throw a WARNING.

```
Usage: check_es_jvm_usage.py [options]

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -v, --verbose
  -H HOST, --host=HOST  The cluster to check
  -P PORT, --port=PORT  The ES port - defaults to 9200
  -C CRITICAL_THRESHOLD, --critical_threshold=CRITICAL_THRESHOLD
                        The level at which we throw a CRITICAL alert -
                        defaults to 97% of the JVM setting
  -W WARNING_THRESHOLD, --warning_threshold=WARNING_THRESHOLD
                        The level at which we throw a WARNING alert - defaults
                        to 85% of the JVM setting


python check_es_unassigned_shards.py --host=myescluster.com
```

###Installation

###Development

###License
This project is licensed under MIT

 [cluster-health]: http://www.elasticsearch.org/guide/reference/api/admin-cluster-health.html
 [node stats]: http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html
 [ElasticSearch]: http://www.elasticsearch.org/


CHANGELOG
2014-07-07 Release 0.1.0
    - initial release of the scripts


check_es_nodes.py
#!/usr/bin/python
from nagioscheck import NagiosCheck, UsageError
from nagioscheck import PerformanceMetric, Status
import urllib2
import optparse

try:
    import json
except ImportError:
    import simplejson as json


class ESNodesCheck(NagiosCheck):

    def __init__(self):

        NagiosCheck.__init__(self)

        self.add_option('E', 'expected_nodes_in_cluster', 'nodes_in_cluster',
                        'This is the expected number of nodes in the cluster')
        self.add_option('H', 'host', 'host', 'The cluster to check')
        self.add_option('P', 'port', 'port', 'The ES port - defaults to 9200')

    def check(self, opts, args):
        host = opts.host
        port = int(opts.port or '9200')
        nodes_in_cluster = int(opts.nodes_in_cluster)

        try:
            response = urllib2.urlopen(r'http://%s:%d/_cluster/health'
                                       % (host, port))
        except urllib2.HTTPError, e:
            raise Status('unknown', ("API failure", None, "API failure:\n\n%s"
                         % str(e)))
        except urllib2.URLError, e:
            raise Status('critical', (e.reason))

        response_body = response.read()

        try:
            es_cluster_health = json.loads(response_body)
        except ValueError:
            raise Status('unknown', ("API returned nonsense",))

        active_cluster_nodes = es_cluster_health['number_of_nodes']

        if active_cluster_nodes != nodes_in_cluster:
            raise Status('CRITICAL', "Number of nodes in the cluster is not %s"
                         % (active_cluster_nodes))
        else:
            raise Status('OK', "Number of nodes in the cluster is as expected")

if __name__ == "__main__":
    ESNodesCheck().run()


check_es_cluster_status.py
#!/usr/bin/python
from nagioscheck import NagiosCheck, UsageError
from nagioscheck import PerformanceMetric, Status
import urllib2
import optparse

try:
    import json
except ImportError:
    import simplejson as json


class ESClusterHealthCheck(NagiosCheck):

    def __init__(self):

        NagiosCheck.__init__(self)

        self.add_option('H', 'host', 'host', 'The cluster to check')
        self.add_option('P', 'port', 'port', 'The ES port - defaults to 9200')

    def check(self, opts, args):
        host = opts.host
        port = int(opts.port or '9200')

        try:
            response = urllib2.urlopen(r'http://%s:%d/_cluster/health'
                                       % (host, port))
        except urllib2.HTTPError, e:
            raise Status('unknown', ("API failure", None,
                         "API failure:\n\n%s" % str(e)))
        except urllib2.URLError, e:
            raise Status('critical', (e.reason))

        response_body = response.read()

        try:
            es_cluster_health = json.loads(response_body)
        except ValueError:
            raise Status('unknown', ("API returned nonsense",))

        cluster_status = es_cluster_health['status'].lower()

        if cluster_status == 'red':
            raise Status("CRITICAL", "Cluster status is currently Red")
        elif cluster_status == 'yellow':
            raise Status("WARNING", "Cluster status is currently Yellow")
        else:
            raise Status("OK", "Cluster status is currently Green")

if __name__ == "__main__":
    ESClusterHealthCheck().run()


check_es_unassigned_shards.py
#!/usr/bin/python
from nagioscheck import NagiosCheck, UsageError
from nagioscheck import PerformanceMetric, Status
import urllib2
import optparse

try:
    import json
except ImportError:
    import simplejson as json

class ESShardsCheck(NagiosCheck):

  def __init__(self):

    NagiosCheck.__init__(self)

    self.add_option('H','host', 'host', 'The cluster to check')
    self.add_option('P','port', 'port', 'The ES port - defaults to 9200')


  def check(self, opts, args):
    host = opts.host
    port = int(opts.port or '9200')

    try:
      response = urllib2.urlopen(r'http://%s:%d/_cluster/health' %(host, port))
    except urllib2.HTTPError, e:
      raise Status('unknown', ("API failure", None, "API failure:\n\n%s" % str(e)))
    except urllib2.URLError, e:
      raise Status('critical', (e.reason))

    response_body = response.read()

    try:
      es_cluster_health = json.loads(response_body)
    except ValueError:
      raise Status('unknown', ("API returned nonsense",))

    unassigned_shards = es_cluster_health['unassigned_shards']

    if es_cluster_health['unassigned_shards'] != unassigned_shards:
      raise Status('CRITICAL', "There are '%s' unassigned shards in the cluster" %unassigned_shards)
    else:
      raise Status('OK', "All shards in the cluster are currently assigned")


if __name__ == "__main__":
  ESShardsCheck().run()


check_es_jvm_usage.py
#!/usr/bin/python
from nagioscheck import NagiosCheck, UsageError
from nagioscheck import PerformanceMetric, Status
import urllib2
import optparse

try:
    import json
except ImportError:
    import simplejson as json


class ESJVMHealthCheck(NagiosCheck):

    def __init__(self):

        NagiosCheck.__init__(self)

        self.add_option('H', 'host', 'host', 'The cluster to check')
        self.add_option('P', 'port', 'port', 'The ES port - defaults to 9200')
        self.add_option('C', 'critical_threshold', 'critical_threshold',
                        'The level at which we throw a CRITICAL alert'
                        ' - defaults to 97% of the JVM setting')
        self.add_option('W', 'warning_threshold', 'warning_threshold',
                        'The level at which we throw a WARNING alert'
                        ' - defaults to 85% of the JVM setting')

    def check(self, opts, args):
        host = opts.host
        port = int(opts.port or '9200')
        critical = int(opts.critical_threshold or '97')
        warning = int(opts.warning_threshold or '85')

        try:
            response = urllib2.urlopen(r'http://%s:%d/_nodes/stats/jvm'
                                       % (host, port))
        except urllib2.HTTPError, e:
            raise Status('unknown', ("API failure", None,
                                     "API failure:\n\n%s" % str(e)))
        except urllib2.URLError, e:
            raise Status('critical', (e.reason))

        response_body = response.read()

        try:
            nodes_jvm_data = json.loads(response_body)
        except ValueError:
            raise Status('unknown', ("API returned nonsense",))

        criticals = 0
        warnings = 0
        msg_details = []

        nodes = nodes_jvm_data['nodes']
        for node in nodes:
            jvm_percentage = nodes[node]['jvm']['mem']['heap_used_percent']
            node_name = nodes[node]['host']
            if int(jvm_percentage) >= critical:
                criticals = criticals + 1
                msg_details.append("Node %s has breached a critical threshold"
                                   % (node_name))
            elif (int(jvm_percentage) >= warning and
                  int(jvm_percentage) < critical):
                warnings = warnings + 1
                msg_details.append("Node %s has breached a warning threshold"
                                   % (node_name))

        if criticals > 0:
            raise Status("Critical",
                         "Cluster status is in critical condition",
                         " ".join(msg_details))
        elif warnings > 0:
            raise Status("Warning", "Cluster health is in a warning state",
                         "\r\n".join(msg_details))
        else:
            raise Status("OK", "Cluster health is good")

if __name__ == "__main__":
    ESJVMHealthCheck().run()
